source/common/upstream/upstream_impl.cc

```diff
@@ -279,6 +279,16 @@ Network::TransportSocketFactory& HostDescriptionImpl::resolveTransportSocketFact
 Host::CreateConnectionData HostImpl::createConnection(
     Event::Dispatcher& dispatcher, const Network::ConnectionSocket::OptionsSharedPtr& options,
     Network::TransportSocketOptionsConstSharedPtr transport_socket_options) const {
+  // If the transport socket options indicate the connection should be
+  // redirected to a proxy, create the TCP connection to the proxy's address not
+  // the host's address.
+  if (transport_socket_options && transport_socket_options->proxyInfo()) {
+    return {createConnection(dispatcher, cluster(),
+                             transport_socket_options->proxyInfo()->proxy_address,
+                             {transport_socket_options->proxyInfo()->proxy_address},
+                             transportSocketFactory(), options, transport_socket_options),
+            shared_from_this()};
+  }
```

Comment:
Up to you but consider having only a single call to createConnection(), but pass in the address/address_list as local variables which default to address() and addressList() and are overridden in the case of the proxy, to make it clear that only the addresses are changing. (Or not, if that doesn't seem cleaner)